File: frontend/src/types/compliance.ts
export type ComplianceStatus = "COMPLIANT" | "WARNING" | "NON_COMPLIANT" | "UNCHECKED" | "SCANNING";
export type RuleType = "keyword_required" | "keyword_forbidden" | "age_limit_days" | "llm_check";
export type Severity = "critical" | "warning";

export interface ComplianceRule {
  id: string;
  name: string;
  description: string | null;
  rule_type: RuleType;
  params: Record<string, unknown>;
  severity: Severity;
  is_active: boolean;
  created_at: string;
  updated_at: string;
}

export interface ComplianceRuleCreate {
  name: string;
  description?: string;
  rule_type: RuleType;
  params: Record<string, unknown>;
  severity: Severity;
}

export interface ComplianceRuleUpdate {
  name?: string;
  description?: string;
  params?: Record<string, unknown>;
  severity?: Severity;
  is_active?: boolean;
}

export interface Location {
  chunk_index: number | null;
  page_number: number | null;
  excerpt: string;
}

export interface ComplianceRuleResult {
  id: string;
  rule_id: string | null;
  rule_name: string;
  rule_type: RuleType;
  severity: Severity;
  passed: boolean;
  detail: string | null;
  locations: Location[] | null;
}

export interface ComplianceReport {
  id: string;
  document_id: string;
  status: ComplianceStatus;
  checked_at: string;
  is_stale: boolean;
  insights: string | null;
  results: ComplianceRuleResult[];
}

export interface ComplianceStats {
  compliant: number;
  warning: number;
  non_compliant: number;
  unchecked: number;
  total_documents: number;
}

export interface ComplianceIssueFailedRule {
  rule_name: string;
  severity: Severity;
  detail: string | null;
}

export interface ComplianceIssueItem {
  document_id: string;
  document_name: string;
  report_id: string;
  status: ComplianceStatus;
  checked_at: string;
  is_stale: boolean;
  failing_rules: ComplianceIssueFailedRule[];
}

export interface ComplianceIssuesResponse {
  items: ComplianceIssueItem[];
  next_cursor: string | null;
  has_more: boolean;
}

File: backend/api_gateway/schemas/compliance.py
from datetime import datetime
from typing import Any, Optional
from uuid import UUID

from pydantic import BaseModel, field_validator


class ComplianceRuleCreate(BaseModel):
    name: str
    description: Optional[str] = None
    rule_type: str
    params: dict[str, Any]
    severity: str

    @field_validator("rule_type")
    @classmethod
    def validate_rule_type(cls, v: str) -> str:
        valid = {"keyword_required", "keyword_forbidden", "age_limit_days", "llm_check"}
        if v not in valid:
            raise ValueError(f"rule_type must be one of {valid}")
        return v

    @field_validator("severity")
    @classmethod
    def validate_severity(cls, v: str) -> str:
        if v not in {"critical", "high", "warning", "info"}:
            raise ValueError("severity must be one of: critical, high, warning, info")
        return v


class ComplianceRuleUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    params: Optional[dict[str, Any]] = None
    severity: Optional[str] = None
    is_active: Optional[bool] = None

    @field_validator("severity")
    @classmethod
    def validate_severity(cls, v: Optional[str]) -> Optional[str]:
        if v is not None and v not in {"critical", "high", "warning", "info"}:
            raise ValueError("severity must be one of: critical, high, warning, info")
        return v


class ComplianceRuleResponse(BaseModel):
    id: UUID
    name: str
    description: Optional[str] = None
    rule_type: str
    params: dict[str, Any]
    severity: str
    is_active: bool
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class LocationResponse(BaseModel):
    chunk_index: Optional[int] = None
    page_number: Optional[int] = None
    excerpt: str


class ComplianceRuleResultResponse(BaseModel):
    id: UUID
    rule_id: Optional[UUID] = None
    rule_name: str
    rule_type: str
    severity: str
    enforcement: str = "advisory"
    passed: bool
    detail: Optional[str] = None
    locations: Optional[list[LocationResponse]] = None

    model_config = {"from_attributes": True}


class ComplianceReportResponse(BaseModel):
    id: UUID
    document_id: UUID
    status: str
    checked_at: datetime
    is_stale: bool = False
    insights: Optional[str] = None
    results: list[ComplianceRuleResultResponse] = []

    model_config = {"from_attributes": True}


class ComplianceStatsResponse(BaseModel):
    compliant: int = 0
    warning: int = 0
    non_compliant: int = 0
    unchecked: int = 0
    total_documents: int = 0


class ComplianceIssueFailedRule(BaseModel):
    rule_name: str
    severity: str
    enforcement: str = "advisory"
    detail: Optional[str] = None


class ComplianceIssueItem(BaseModel):
    document_id: UUID
    document_name: str
    report_id: UUID
    status: str
    checked_at: datetime
    is_stale: bool = False
    failing_rules: list[ComplianceIssueFailedRule] = []


class ComplianceIssuesResponse(BaseModel):
    items: list[ComplianceIssueItem]
    next_cursor: Optional[str] = None
    has_more: bool = False
